C error on comparison fixed array with aliased fixed array with operator overloading

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 5ec9bb5.fbc3fc6
OS: linux, Linux version 6.6.8-arch1-1 (linux@archlinux) (gcc (GCC) 13.2.1 20230801, GNU ld (GNU Binutils) 2.41.0) #1 SMP PREEMPT_DYNAMIC Thu, 21 Dec 2023 19:01:01 +0000
Processor: 16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P

getwd: /home/ge/Code/v/netaddr/fixed_array_issues
vexe: /home/ge/.vlang/v
vexe mtime: 2024-11-20 20:10:01

vroot: OK, value: /home/ge/.vlang
VMODULES: OK, value: /home/ge/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.46.1
Git vroot status: weekly.2024.47-2-gfbc3fc6f
.git/config present: true

CC version: cc (GCC) 14.2.1 20240910
emcc version: N/A
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg fixed_array_alias_eq.v`

```v
module main

import encoding.binary

pub type Addr = [4]u8

pub fn (a Addr) u32() u32 {
	return binary.big_endian_u32_fixed(a)
}

pub fn (a Addr) == (oth Addr) bool {
	return a.u32() == oth.u32()
}

pub fn (a Addr) < (oth Addr) bool {
	return a.u32() < oth.u32()
}

fn main() {
	addr := Addr([u8(127), 0, 0, 1]!)
	assert addr == Addr([u8(127), 0, 0, 1]!) // OK
	assert addr == [u8(127), 0, 0, 1]! // C error
	assert addr < Addr([u8(127), 0, 0, 2]!) // OK
	assert addr < [u8(127), 0, 0, 2]! // C error
}
```

**What did you expect to see?**

Assertion passed, no C errors.

**What did you see instead?**

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/fixed_array_alias_eq.01JD5PSY5JJ0J5NDSAVSDC8PV9.tmp.c:14368: error: expression expected before '{'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzNlNTFiNGU3Y2M1YTc4NTQyZTY5ZTYiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.i4RWTFSS49S9Ko0WtCKhwSoQRxgcb1aTI_8Ne3jfRbw">Huly&reg;: <b>V_0.6-21367</b></a></sub>

--- a/vlib/v/gen/c/infix.v
+++ b/vlib/v/gen/c/infix.v
@@ -165,12 +165,20 @@ fn (mut g Gen) infix_expr_eq_op(node ast.InfixExpr) {
 		if eq_operator_expects_ptr {
 			g.write('&')
 		}
-		g.expr(node.left)
+		if node.left is ast.ArrayInit && g.table.sym(node.left_type).kind == .array_fixed {
+			g.fixed_array_init_with_cast(node.left, node.left_type)
+		} else {
+			g.expr(node.left)
+		}
 		g.write2(', ', '*'.repeat(right.typ.nr_muls()))
 		if eq_operator_expects_ptr {
 			g.write('&')
 		}
-		g.expr(node.right)
+		if node.right is ast.ArrayInit && g.table.sym(node.right_type).kind == .array_fixed {
+			g.fixed_array_init_with_cast(node.right, node.right_type)
+		} else {
+			g.expr(node.right)
+		}
 		g.write(')')
 	} else if left.unaliased.idx() == right.unaliased.idx()
 		&& left.sym.kind in [.array, .array_fixed, .alias, .map, .struct, .sum_type, .interface] {
@@ -414,7 +422,7 @@ fn (mut g Gen) infix_expr_cmp_op(node ast.InfixExpr) {
 			g.expr(node.left)
 			g.write(')')
 		}
-	} else if left.sym.kind == right.sym.kind && has_operator_overloading {
+	} else if left.unaliased_sym.kind == right.unaliased_sym.kind && has_operator_overloading {
 		if node.op in [.le, .ge] {
 			g.write('!')
 		}
@@ -424,12 +432,20 @@ fn (mut g Gen) infix_expr_cmp_op(node ast.InfixExpr) {
 			if operator_expects_ptr {
 				g.write('&')
 			}
-			g.expr(node.left)
+			if node.left is ast.ArrayInit && g.table.sym(node.left_type).kind == .array_fixed {
+				g.fixed_array_init_with_cast(node.left, node.left_type)
+			} else {
+				g.expr(node.left)
+			}
 			g.write2(', ', '*'.repeat(right.typ.nr_muls()))
 			if operator_expects_ptr {
 				g.write('&')
 			}
-			g.expr(node.right)
+			if node.right is ast.ArrayInit && g.table.sym(node.right_type).kind == .array_fixed {
+				g.fixed_array_init_with_cast(node.right, node.right_type)
+			} else {
+				g.expr(node.right)
+			}
 			g.write(')')
 		} else {
 			g.write2('(', '*'.repeat(right.typ.nr_muls()))

--- a/vlib/v/tests/aliases/alias_fixed_array_infix_expr_test.v
+++ b/vlib/v/tests/aliases/alias_fixed_array_infix_expr_test.v
@@ -1,0 +1,28 @@
+import encoding.binary
+
+pub type Addr = [4]u8
+
+pub fn (a Addr) u32() u32 {
+	return binary.big_endian_u32_fixed(a)
+}
+
+pub fn (a Addr) == (oth Addr) bool {
+	return a.u32() == oth.u32()
+}
+
+pub fn (a Addr) < (oth Addr) bool {
+	return a.u32() < oth.u32()
+}
+
+fn test_alias_fixed_array_infix_expr() {
+	addr := Addr([u8(127), 0, 0, 1]!)
+
+	assert addr == Addr([u8(127), 0, 0, 1]!)
+	assert Addr([u8(127), 0, 0, 1]!) == addr
+
+	assert addr == [u8(127), 0, 0, 1]!
+	assert [u8(127), 0, 0, 1]! == addr
+
+	assert addr < Addr([u8(127), 0, 0, 2]!)
+	assert addr < [u8(127), 0, 0, 2]!
+}

